### PR TITLE
Mobile2.0 .net app

### PR DESCRIPTION
**Pull request by hrajput89**
#### Fixes #.

- [x] Ran `flake8`
- [x] Ran `run_repo_tests.sh`

#### Changes proposed in this pull request:

- .net App for mobile2.0

### CODE REVIEW

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DataSourceMethods.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace Couchbase.Lite.Testing
+{
+    internal static class DataSourceMethods
+    {
+        public static void Database([NotNull] NameValueCollection args,
+                                    [NotNull] IReadOnlyDictionary<string, object> postBody,
+                                    [NotNull] HttpListenerResponse response)
+        {
+            With<Database>(postBody, "database", db => response.WriteEmptyBody());
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DataSourceMethods.cs]: I guess this was just an experimental file. Will remove it.

[the remark this answers is not in the document]
Comment (hrajput89) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DataSourceMethods.cs]: I guess this was just an experimental file. Removed it.

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DatabaseMethods.cs [hunk 2]
```diff
@@ -0,0 +1,511 @@
...
+            var handle = postBody[key].ToString();
+            var db = MemoryMap.Get<T>(handle);
+            await action(db);
+        }
+
+        #endregion
+
+        #region Internal Methods
+
+        //internal static void DatabaseAddChangeListener([NotNull] NameValueCollection args,
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DatabaseMethods.cs]: We need this function. I remember that I was not able to write it correctly hence I commented it. @raghusarangapani, if you think we don't need this function, I'll remove it.

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DatabaseMethods.cs [hunk 3]
```diff
@@ -0,0 +1,511 @@
...
+            With<Database>(postBody, "database", db => db.Close());
+            response.WriteEmptyBody();
+        }
+
+        internal static void DatabaseCompact([NotNull] NameValueCollection args,
+                                             [NotNull] IReadOnlyDictionary<string, object> postBody,
+                                             [NotNull] HttpListenerResponse response)
+        {
+            With<Database>(postBody, "database", db => db.Compact());
+            int bodyObj = -1;
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DatabaseMethods.cs]: done

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DatabaseMethods.cs [hunk 4]
```diff
@@ -0,0 +1,511 @@
...
+                response.WriteBody(databaseId);
+            }
+        }
+
+        internal static void DatabaseDeleteDB([NotNull] NameValueCollection args,
+            [NotNull] IReadOnlyDictionary<string, object> postBody,
+            [NotNull] HttpListenerResponse response)
+        {
+            With<Database>(postBody, "database", db => db.Delete());
+            int bodyResponse = -1;
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DatabaseMethods.cs]: done

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DatabaseMethods.cs [hunk 5]
```diff
@@ -0,0 +1,511 @@
...
+            [NotNull] IReadOnlyDictionary<string, object> postBody,
+            [NotNull] HttpListenerResponse response)
+        {
+            string docId = null;
+            if (postBody.ContainsKey("id"))
+            {
+                docId = postBody["id"].ToString();
+            }
+
+            With<Database>(postBody, "database", db =>
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DatabaseMethods.cs]: I think it is intentional, as Database should throw an error for null docId.

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DatabaseMethods.cs [hunk 6]
```diff
@@ -0,0 +1,511 @@
...
+                    } catch (InvalidOperationException e)
+                    {
+                        Console.WriteLine("Error deleting DB" + e.Message);
+                        response.WriteBody(e.Message);
+                    }
+                });
+            });
+        }
+
+        //internal static void DatabaseRemoveChangeListener([NotNull] NameValueCollection args,
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DatabaseMethods.cs]: Was not able to wrote correct implementation. Hence commented it out. We need to write it properly.

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DatabaseMethods.cs [hunk 7]
```diff
@@ -0,0 +1,511 @@
...
+            {
+                db.InBatch(() =>
+               {
+                    foreach (var body in docDict)
+                    {
+                        string id = body.Key;
+                        var docBody = (Dictionary<string, object>)body.Value;
+                        docBody.Remove("_id");
+                        MutableDocument doc = new MutableDocument(id, docBody);
+                       db.Save(doc);
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DatabaseMethods.cs]: fixed

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DocumentMethods.cs [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+            var key = postBody["key"].ToString();
+            With<MutableDocument>(postBody, "document", doc => response.WriteBody(doc.Contains(key)));
+        }
+
+        public static void DocumentDelete([NotNull] NameValueCollection args,
+            [NotNull] IReadOnlyDictionary<string, object> postBody,
+            [NotNull] HttpListenerResponse response)
+        {
+            With<Database>(postBody, "database", db => With<Document>(postBody, "document", db.Delete));
+            int bodyResponse = -1;
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DocumentMethods.cs]: done

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/HttpListenerResponseExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+                response.Close();
+            } catch (ObjectDisposedException) {
+                // Swallow...other side closed the connection
+            }
+        }
+
+        public static void WriteEmptyBody([NotNull]this HttpListenerResponse response, HttpStatusCode code = HttpStatusCode.OK)
+        {
+            try {
+				var body = Encoding.UTF8.GetBytes("I-1");
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/HttpListenerResponseExtensions.cs]: fixed

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/QueryMethods.cs [hunk 10]
```diff
@@ -0,0 +1,940 @@
...
+
+        internal static void QueryDistinct([NotNull] NameValueCollection args,
+            [NotNull] IReadOnlyDictionary<string, object> postBody,
+            [NotNull] HttpListenerResponse response)
+        {
+            With<ISelectResult>(postBody, "select_result", se =>
+            {
+                With<IDataSource>(postBody, "from_pop", ds =>
+                {
+                    With<IExpression>(postBody, "whr_key_prop", exp => response.WriteBody(MemoryMap.Store(QueryBuilder.Select(se).From(ds).Where(exp))));
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/QueryMethods.cs]: Fixed for Distinct.

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/QueryMethods.cs [hunk 11]
```diff
@@ -0,0 +1,940 @@
...
+            [NotNull] HttpListenerResponse response)
+        {
+            With<IQuery>(postBody, "query", query => response.WriteBody(MemoryMap.Store(query.Execute())));
+        }
+
+        internal static void QueryNextResult([NotNull] NameValueCollection args,
+            [NotNull] IReadOnlyDictionary<string, object> postBody,
+            [NotNull] HttpListenerResponse response)
+        {
+            With<ISelectResult>(postBody, "select_result", se => response.WriteBody(MemoryMap.Store(QueryBuilder.Select(se))));
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/QueryMethods.cs]: Yes, it is same in all other platforms. However, we don't have any case where we are using this method. So may be we can remove this code. @raghusarangapani and @sridevi-15, do let me know your opinion.

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/QueryMethods.cs [hunk 12]
```diff
@@ -0,0 +1,940 @@
...
+                       .Select(
+                           SelectResult.Expression(Expression.Property(prop1).From(main)),
+                           SelectResult.Expression(Expression.Property(prop2).From(main)),
+                           SelectResult.Expression(Expression.Property(prop3).From(secondary)))
+                       .From(DataSource.Database(db).As(main))
+                       .Join(Join.InnerJoin(DataSource.Database(db).As(secondary))
+                             .On(Expression.Property(joinKey1).From(secondary).EqualTo(Expression.Property(joinKey2).From(main))
+                                 .And(Expression.Property(whrKey1).From(secondary).EqualTo(Expression.String(whrVal1)))
+                                 .And(Expression.Property(whrKey2).From(main).EqualTo(Expression.Int(whrVal2)))))
+                       //.OrderBy(Ordering.Expression(Expression.Property(prop1).From(main)).Ascending())
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/QueryMethods.cs]: done.

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/QueryMethods.cs [hunk 13]
```diff
@@ -0,0 +1,940 @@
...
+                using (IQuery query = QueryBuilder
+                       .Select(
+                           SelectResult.Expression(Expression.Property(prop1).From(main)).As(firstName),
+                           SelectResult.Expression(Expression.Property(prop1).From(secondary)).As(secondName),
+                           SelectResult.Expression(Expression.Property(prop2).From(secondary)))
+                       .From(DataSource.Database(db).As(main))
+                       .Join(Join.CrossJoin(DataSource.Database(db).As(secondary)))
+                       .Where(Expression.Property(whrKey1).From(main).EqualTo(Expression.String(whrVal1))
+                              .And(Expression.Property(whrKey2).From(secondary).EqualTo(Expression.String(whrVal2))))
+                       //.OrderBy(Ordering.Expression(Expression.Property(prop1).From(main)).Ascending())
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/QueryMethods.cs]: done

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/ReplicatorConfigurationMethods.cs [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+            {
+                replType = ReplicatorType.PushAndPull;
+            }
+            With<ReplicatorConfiguration>(postBody, "configuration", repConf =>
+            {
+                repConf.ReplicatorType = replType;
+            });
+        }
+
+        //public static void setPinnedServerCertificate([NotNull] NameValueCollection args,
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/ReplicatorConfigurationMethods.cs]: Uncommented and modified according to changes in Android app code from master branch.

**Comment by hrajput89** [comment]
Made changes requested by @raghusarangapani. We need to write ChangeListerner methods properly. I haven't uncommented them yet.

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DataSourceMethods.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (raghusarangapani) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DataSourceMethods.cs]: This file exists in other platforms

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DatabaseMethods.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (raghusarangapani) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DatabaseMethods.cs]: I am not sure, @sridevi-15 what do you think?

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/TestServer.cs [hunk 15]
```diff
@@ -33,7 +33,7 @@ public sealed class TestServer
     {
         #region Constants
 
-        private const ushort Port = 8080;
+        private const ushort Port = 55555;
```
Comment (raghusarangapani) [inline CBLClient/Apps/CBLTestServer-Dotnet/TestServer/TestServer.cs]: Can you change the port to 8080?

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/HttpListenerResponseExtensions.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (raghusarangapani) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/HttpListenerResponseExtensions.cs]: Did you forget to push, I don't see this in your latest commit

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DatabaseMethods.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (hrajput89) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DatabaseMethods.cs]: added Todo comment

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DatabaseMethods.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (hrajput89) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DatabaseMethods.cs]: removed it as the we should get exception for null doc.

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DatabaseMethods.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (hrajput89) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DatabaseMethods.cs]: added todo comment

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer.UWP/Properties/AssemblyInfo.cs [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+// Version information for an assembly consists of the following four values:
+//
+//      Major Version
+//      Minor Version 
+//      Build Number
+//      Revision
+//
+// You can specify all the values or you can default the Build and Revision Numbers 
+// by using the '*' as shown below:
+// [assembly: AssemblyVersion("1.0.*")]
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer.UWP/Properties/AssemblyInfo.cs]: removed the complete comment section.

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DatabaseMethods.cs [hunk 17]
```diff
@@ -0,0 +1,511 @@
...
+                    foreach (var docId in docIds)
+                    {
+                        Document doc = db.GetDocument(docId);
+                        db.Delete(doc);
+                    }
+                });
+            });
+            response.WriteEmptyBody();
+        }
+        // TODO
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DatabaseMethods.cs]: In android the method code was different. The current code was same as deleteDB. So I changed the method definition to what is there in android.

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer.NetCore/QueryMethods.cs [hunk 18]
```diff
@@ -0,0 +1,147 @@
+////
```
[the remark this answers is not in the document]
Comment (hrajput89) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer.NetCore/QueryMethods.cs]: removed the file

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DataSourceMethods.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (sridevi-15) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DataSourceMethods.cs]: I still do not see this change

[the remark this answers is not in the document]
Comment (sridevi-15) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DataSourceMethods.cs]: looks like this is removed. According to Raghu, this should be there

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/ReplicatorConfigurationMethods.cs [hunk 19]
```diff
@@ -0,0 +1,309 @@
...
+                replType = ReplicatorType.PushAndPull;
+            }
+            With<ReplicatorConfiguration>(postBody, "configuration", repConf =>
+            {
+                repConf.ReplicatorType = replType;
+            });
+            response.WriteEmptyBody();
+        }
+
+        // TODO have to fix this method
```
Comment (sridevi-15) [inline CBLClient/Apps/CBLTestServer-Dotnet/TestServer/ReplicatorConfigurationMethods.cs]: I hope this method fix with your latest changes. If yes, do you want to remove this TODO?

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer.UWP/Properties/AssemblyInfo.cs (hunk 16 above)
[the remark this answers is not in the document]
Comment (sridevi-15) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer.UWP/Properties/AssemblyInfo.cs]: I do not see this change

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/ReplicatorConfigurationMethods.cs [hunk 20]
```diff
@@ -286,14 +287,22 @@ public class ReplicatorConfigurationMethods
...
+            var obj = postBody["cert"];
+            using (MemoryStream ms = new MemoryStream()){
+                binaryFormatter.Serialize(ms, obj);
+                 certs = ms.ToArray();
+
+            }
             With<ReplicatorConfiguration>(postBody, "configuration", repConf =>
             {
-                repConf.PinnedServerCertificate = cert;
+                repConf.PinnedServerCertificate = null;
```
Comment (sridevi-15) [inline CBLClient/Apps/CBLTestServer-Dotnet/TestServer/ReplicatorConfigurationMethods.cs]: it is assigning to null?

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/ReplicatorConfigurationMethods.cs (hunk 19 above)
Comment (hrajput89) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/ReplicatorConfigurationMethods.cs]: No, I haven't fixed this method, as I was not able to find the way to convert var to byte[].

File: CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DataSourceMethods.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (raghusarangapani) [reply CBLClient/Apps/CBLTestServer-Dotnet/TestServer/DataSourceMethods.cs]: Was this resolved?

**Pull request closed without merging**